src/gpio.rs

```diff
@@ -109,18 +121,57 @@ impl From<bool> for GpioState {
 pub struct Gpios<'a> {
     num_gpios: usize,
     curr_gpio: usize,
+    used_pins: [u8; MAX_NUM_PINS_DIV_8],
     lifetime: PhantomData<&'a ()>,
 }
 
+impl<'a> Gpios<'a> {
+    fn pin_used(&self, pin_number: usize) -> bool {
+        self.used_pins[pin_number / 8] & 1 << (pin_number % 8) != 0
+    }
+
+    fn set_pin_used(&mut self, pin_number: usize) {
+        self.used_pins[pin_number / 8] |= 1 << (pin_number % 8);
+    }
+
+    pub fn next_at(&mut self, pinid: usize) -> Option<Gpio<'a>> {
+        if let Ok(gpio_num) =
+            syscalls::command(DRIVER_NUMBER, command_nr::GET_PIN_BY_PINID, pinid, 0)
+        {
+            if !self.pin_used(gpio_num) && gpio_num < self.num_gpios {
+                let item = Gpio {
+                    gpio_num: self.curr_gpio,
+                    lifetime: PhantomData,
+                };
+                self.set_pin_used(gpio_num);
+                self.curr_gpio = gpio_num + 1;
+                Some(item)
+            } else {
+                None
+            }
+        } else {
+            None
+        }
+    }
+
+    pub fn reset(&mut self) {
```

Comment:
Wouldn't this reintroduce the soundness problem mentioned before?

Reply:
If we keep track of every used pin, no, otherwise yes.